core/domain/collection_services_test.py

```diff
@@ -116,6 +117,180 @@ def test_get_collection_titles_and_categories(self):
                 })
 
 
+class ExplorationInCollectionContextUnitTests(CollectionServicesUnitTests):
+    """Tests query and recording methods related to explorations which are
+    played in the context of a collection.
+    """
+
+    COL_ID_0 = '0_collection_id'
+    COL_ID_1 = '1_collection_id'
+    EXP_ID_0 = '0_exploration_id'
+    EXP_ID_1 = '1_exploration_id'
+    EXP_ID_2 = '2_exploration_id'
+
+    def _get_completion_model(self, user_id, collection_id):
+        return user_models.ExplorationInCollectionCompletionModel.get(
+            user_id, collection_id)
+
+    def _record_completion(self, user_id, collection_id, exploration_id):
+        collection_services.record_played_exploration_in_collection_context(
+            user_id, collection_id, exploration_id)
+
+    def setUp(self):
+        super(ExplorationInCollectionContextUnitTests, self).setUp()
+
+        # Create a new collection and exploration.
+        self.save_new_valid_collection(
+            self.COL_ID_0, self.OWNER_ID, exploration_id=self.EXP_ID_0)
+
+        # Create another two explorations and add them to the collection.
+        for exp_id in [self.EXP_ID_1, self.EXP_ID_2]:
+            self.save_new_valid_exploration(exp_id, self.OWNER_ID)
+            collection_services.update_collection(
+                self.OWNER_ID, self.COL_ID_0, [{
+                    'cmd': collection_domain.CMD_ADD_COLLECTION_NODE,
+                    'exploration_id': exp_id
+                }], 'Added new exploration')
+
+    def test_get_completed_exploration_ids(self):
+        # There should be no exception if the user or collection do not exist.
```

Comment:
Clarified the return part. I'm indifferent about the exceptions; they could be triggered (if present) if the learner collection data handler is called for a non-existent collection ID. Do you have a preference? Also, the policy here is 'no model = no progress yet', so we assume that means the collection exists. We only create an instance of the model if progress has been made, so all other states are "not yet started."

Regarding your other question, nothing currently happens if the collection is deleted. I'm also not quite sure how to approach this problem; the progress model is keyed on both a user ID and collection ID, so it's not straightforward to prune them. This might warrant a mapreduce job which checks all progress models, pulls out their collection IDs, and checks them against the current list of collection IDs.